quarkc/versioning.py

```diff
@@ -1,55 +1,96 @@
-import semantic_version as semver
+from collections import namedtuple
+from re import compile as re
 
 from ._metadata import __version__
 from .helpers import lineinfo
 import ast
 import messages
+from messages import Error, Warning
 
 
-COMPILER_VERSION = semver.Version(__version__)
+class Version:
 
+    class Full(namedtuple('FullVersion', ['major', 'minor', 'patch'])):
+        """Version of form `X.Y.Z`."""
 
-def version_spec_string_errors(spec_string, location):
-    try:
-        spec = semver.Spec(spec_string)
-    except ValueError:
-        yield messages.version_spec_malformed(
+        def is_compatible_with(self, compiler_version):
+            return (compiler_version.major == self.major and
+                    (compiler_version.minor > self.minor or
+                     (compiler_version.minor == self.minor and
+                      (compiler_version.patch >= self.patch))))
+
+        def __str__(self):
+            return "%s.%s.%s" % (self.major, self.minor, self.patch)
+
+    class Star(object):
+        """Version `*` singleton."""
+
+        @staticmethod
+        def is_compatible_with(_):
+            return True
+
+    class Invalid(object):
+        """Version that didn't parse correctly."""
+
+    _re = re("^(\d)\.(\d)(?:\.(\d))?|(\*)$")
+
+    @classmethod
+    def parse(cls, source):
+        match = cls._re.match(source)
+        if not match:
+            return cls.Invalid
+        major, minor, patch, star = match.groups()
+        if star:
+            return cls.Star
```

Comment:
I see, I'll find/make something like the Ruby Singleton during slack time, so we can avoid this. You have valid point.